### ISSUE TITLE
Grid Subsampling

### ISSUE THREAD

**Issue by edshkim98** [HuguesTHOMAS/KPConv-PyTorch#78]
Hi, @HuguesTHOMAS !
Thank you for your code.
I have several questions regarding the grid subsampling strategy you used.
1. I cant seem to find much information about grid subsampling so it will be great if you could explain the technique or provide me a link for the explanation of the subsampling.
2. Did you use grid sampling to preprocess the raw points originally and also to downsample points in every convolution layer?
3. Finally, how many points are sampled on S3DIS data initially and then fed into the network?
Thank you!

**Comment by HuguesTHOMAS** [comment]
1. Grid subsmapling is pretty basic, you just keep one point per voxel in a 3D grid. You can have a look at my [PhD thesis](https://pastel.archives-ouvertes.fr/tel-02458455/document) (part II.4.) for more details.
2. Yes and yes
3. The first subsampling of S3DIS point cloud is done using the value `first_subsampling_dl`. Then the subpart of the clouds that are fed to the network are spherical subclouds controlled by the parameter `in_radius`. See part IV.2.8.c of my thesis for details about the input spheres

**Comment by edshkim98** [comment]
@HuguesTHOMAS  Thank you for the reply and providing me your thesis! I will make sure to read it. Now I understand the basic of Grid subsampling, but there are still some confusions. 
1. If there are points with different labels within a single voxel, how does the function label the new barycentric point?  
2. Also, for the training S3DIS dataset, you set first_subsampling_dl value to sample input points to the network. Is this the only necessary step for preprocessing the S3DIS data except normalization etc? Then. how did you preprocess for validation&testing dataset? 
Thank you!!

**Comment by HuguesTHOMAS** [comment]
@edshkim98,

1. The label is chosen by a vote (the most represented label in the voxel is kept).
2. Yes it is the only preprocessing I do, and I use the same preprocessing for training validation and test sets. For test and validation, the predictions of the network (which are computed on subsampled point clouds) are reprojected to the original point clouds via a simple nearest-neighbor interpolation. This is mandatory if you want to compare the results to other methods

**Comment by edshkim98** [comment]
@HuguesTHOMAS Thank you for the reply. 
1. When you say use a simple knn interpolation, did you mean inferencing on the full point cloud data? If so, wouldn't that lead to decrease in performance significantly since the subsampled points are much smaller compare to the original points?
2. Do you remember how much the performance was affected by the interpolation during the testing?
3. Could there be another method to inference on the full point cloud data?
Thank you for the help!

**Comment by HuguesTHOMAS** [comment]
1. When I say "nearest-neighbor interpolation", I mean nearest-neighbor interpolation, The prediction in one point of the original point cloud is equal to the prediction of the nearest point in the subsampled point cloud (the inference is done on the subsampled point clouds).
2. Density in S3DIS dataset is pretty regular, so the interpolation nearly did not affect the performances. (class mIoU were nearly equal on the subsampled clouds and on the original clouds)
3. There could be other methods. Any interpolation function would do actually, but you would not see much difference in the final score in my opinion.

**Comment by edshkim98** [comment]
@HuguesTHOMAS Hi, thank you for the reply.
I have a slight different question regarding visualizing feature map.
In you KPConv paper, you have visualized feature maps for different layers (figure 6) and also effective receptive field (figure 7).
Could you please let me know how did you visualize this feature map?
I am currently using Pytorch.
Thank you very much in advance.

**Comment by HuguesTHOMAS** [comment]
The first maps from fig 6 are not hard to visualize. Just perform a forward pass of the network , take the features of the layer you want to visualize. So here:
https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/bb0ea445b168463d4b7dffee8b61c235775fc33a/models/architectures.py#L329-L332

keep the values as a variable with something like  

```
if block_i == visu_ind_you_chose
    self.features_to_visualize = x.cpu().detach().numpy()
```

then after the forward pass is done you can save these value wherever you want, reproject them on the pointcloud of the first layer for nicer visualization (use the points corresponding to the right layer stored in `batch.points`).

**Comment by HuguesTHOMAS** [comment]
As for the ERF in fig 7, this is more complex, you have to perform a forward pass, then fake a gradient (change of value) at a particular point for the feature of a certain layer. Then backpropagate this gradient to the features of the first layer. But to achieve that, the parameters of your networks need to be fixed, and the features of the first layer need to be trainable. I never did that in PyTorch so I cannot really help you, but you can find more details in the Tensorflow implementation.

**Comment by anandrajaraman** [comment]
@HuguesTHOMAS Hi, I wanted to get some further clarifications on visualizing the feature maps (as in Fig 6). Can you provide some more details (or a small example snippet) about this aspect:
> reproject them on the pointcloud of the first layer for nicer visualization (use the points corresponding to the right layer stored in batch.points
I am trying to visualize the activations on a complete pointcloud scan of SemanticKITTI. Since we operate on subsampled grids, I want to understand how these feature maps will map back to the original point cloud and colored by the activations (to be able visualize how the learning progresses from the initial to the later layers). 
Thank you very much.

**Comment by HuguesTHOMAS** [comment]
You can see the code of the original tensorflow implementation:
https://github.com/HuguesTHOMAS/KPConv/blob/16bfbb96ac9af48a3c829d1f8123152d35b63862/utils/visualizer.py#L86
and 
https://github.com/HuguesTHOMAS/KPConv/blob/16bfbb96ac9af48a3c829d1f8123152d35b63862/utils/visualizer.py#L248

More specifically, the snippet you want is here:
https://github.com/HuguesTHOMAS/KPConv/blob/16bfbb96ac9af48a3c829d1f8123152d35b63862/utils/visualizer.py#L293-L302

This code projects the values from the subsampled points of layer X, to the subsampled points of the first layer. Then you can use the test reprojection indices to project from the first layer to the original data points. like it is done for validation/test:
https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/7fefb6a8d38fd304775199777ad01d9f1546e2ff/utils/trainer.py#L734-L735

**Comment by anandrajaraman** [comment]
@HuguesTHOMAS Thank you very much for the response. I had started looking at the tensorflow implementation as you have suggested here. I want to confirm the following aspects
1) I understand the `features` we would use in the equivalent pytorch version of this tf implementation snippet (https://github.com/HuguesTHOMAS/KPConv/blob/16bfbb96ac9af48a3c829d1f8123152d35b63862/utils/visualizer.py#L248) would be what you described above  
```
if block_i == visu_ind_you_chose:
    self.features_to_visualize = x.cpu().detach().numpy()
```
and the `layer_idx` from the tf implementation would be the `vis_ind_you_chose`. Is my understanding correct?

2)  I think you don't specify the `density_parameter` in the configs for the pytorch implementation which is needed to compute the `radius` here (https://github.com/HuguesTHOMAS/KPConv/blob/16bfbb96ac9af48a3c829d1f8123152d35b63862/utils/visualizer.py#L120). How do you recommend getting this value for the pytorch configs?

3) Since there are some subtle differences in how the data is accessed in pytorch vs how they are done in the tf implementation,  I think this snippet from pytorch (https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/master/utils/visualizer.py#L166-L186) is the relevant one for getting the necessary data to be passed to `update_top_activations()` in pytorch. I wanted to confirm I am accessing the right containers and not have any mismatch. 

Thank you for your help.

**Comment by HuguesTHOMAS** [comment]
1. Yes indeed, but be careful of something, what I generally call `layer_idx`, is different from what I call `block_idx`. In my architecture, I have different subsampling levels, which I call layers, and in each layer, I have some blocks, (for ex the resnet convolutions). So `layer_idx` refers to the level of subsampling and I generally have 5 of them. There are more `block_idx` as we have more than one block per layer.

2. I changed the name of the parameters in this PyTorch version for clarity, `density_parameter` is now called `conv_radius`. A good place to understand layers, blocks and convolution radius is the KPCNN class init method:
https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/7fefb6a8d38fd304775199777ad01d9f1546e2ff/models/architectures.py#L201-L203

3. Yes you are right

**Comment by pixar0407** [comment]
> 1. Grid subsmapling is pretty basic, you just keep one point per voxel in a 3D grid. You can have a look at my [PhD thesis](https://pastel.archives-ouvertes.fr/tel-02458455/document) (part II.4.) for more details.
> 2. Yes and yes
> 3. The first subsampling of S3DIS point cloud is done using the value `first_subsampling_dl`. Then the subpart of the clouds that are fed to the network are spherical subclouds controlled by the parameter `in_radius`. See part IV.2.8.c of my thesis for details about the input spheres

@HuguesTHOMAS 

Hello, again. 
Your PhD thesis seems really helpful to me. Thanks for good paper.
But, i want to clarify things and ask about them.

1.  Is part II.4. of your thesis only about the pre-process stage? specifically, when we have raw data, we sub-sample it before feeding them to a neural network(like KPconv-net). **Or** is part II.4. about the sampling procedure between layers of a whole neural networks? specifically, a lot of neural network (like KPconv-net, pointnet++, RandLA-net etc) down-samples the points between layers and up-samples them between layers for segmentation. 

2. At the 3.3 section of KPconv paper (**not your phd thesis**), 'Subsampling to deal with varying densities" is about pre-processing for raw dataset Or about sampling between layers?

3. About the sampling procedure between layers, I can't find the sampling-between-layer part among the code. 
i have two guesses.

3-1. i guess that the 'batch' variable in Line 143 already has points which is sampled per layer at pre-processing stage. 

And at the (https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/7fefb6a8d38fd304775199777ad01d9f1546e2ff/models/blocks.py#L563), it seems that by configuring "layer_ind", we may get the pre-sampled( at layer level, not raw data level) points(?).

if this is true, then why didn't implement your code as other people? Other people used to down-sample points at each layer of the network. 

3-2. **Or** does the sampling procedure is done by Pooling layer(https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/7fefb6a8d38fd304775199777ad01d9f1546e2ff/models/blocks.py#L643)?

But, I think that max_pooling function only gives max_pooled features, not points' xyz values.


this is end of my question. thank you.

**Comment by HuguesTHOMAS** [comment]
Hi @pixar0407,

1. All subsamplings in my work are done with a grid, whether it is the initial input preprocessing or the layer to layer subsampling.

2. It is about the initial pre-processing step, which controls the density of points that go into the network. The subsequent subsamplings are here to reduce this controlled density from layer to layer.

3. Exactly, we perform several operations during the preprocessing step to reduce the load on the GPU computations. Subsampling for each layer is one of them, in addition to finding the neighbors for every convolution location at each layer. Why do we do that? Because computing neighborhoods in point clouds is not easily done on GPU. We have the means to do it extremely fast on CPU (with KDTree acceleration, see nanoflann library). Furthermore, the CPU neighbor computations can be parallelized on multiple processes and done while the network GPU computations are performed. This is a huge performance gain. Now, about the grid subsampling, it could be implemented and be very fast on GPU, but we actually need the subsampling of all layers to be done to be able to also compute the neighbors, so we have to do it with the neighbor computations on CPU (it is very fast anyway).

4. Not really we are effectively using a "strided convolution" meaning a convolution happening only on a subsampled version of the layer input structure (one pixel over 2 for images, grid subsampled points for point clouds). If you look more closely at the code, the maxpoolling that you see is just performed on the resnet shortcut features, because they also need to be pooled to the new point locations and no convolution is performed on them.

Hope that clarifies things for you!
Best,
Hugues

**Comment by pixar0407** [comment]
Hi, @HuguesTHOMAS 

Oh! Thank you for the detailed explanation.
I will go deep into your quality code :)

hope you being good!

**Comment by HuguesTHOMAS** [comment]
Sorry I did not see your question at the time.

s_ for support and q_for query, pretty standard for point cloud neighbor computations
